docs/LoongArch-ELF-ABI-EN.adoc

```diff
@@ -628,3 +628,197 @@ with check 32-bit unsigned overflow
 |/lib32/ld-linux-loongarch-ilp32s.so.1
 |===
+
+== Procedure Calling Convention
+
+The basic principle of the LoongArch procedure calling convention is to pass arguments in registers as much as possible (i.e. floating-point arguments are passed in floating-point registers and non floating-point arguments are passed in general-purpose registers, as much as possible); arguments are passed on the stack only when no appropriate register is available.
+
+The argument registers are:
+
+. Eight floating-point registers `fa0-fa7`, and `fa0-fa1` are also used to return values. They are all used to pass floating-point arguments.
+
+. Eight general-purpose registers `a0-a7`, with `a0-a1` reused to return values.
+
+Generally, the GARs are used to pass fixed-point arguments, and floating-point arguments when no FAR is available.
+Bit fields are stored in little endian.
+In addition, subroutines should ensure that the values of general-purpose registers `s0-s9` and floating-point registers `fs0-fs7` are preserved across procedure calls.
+
+==== Note
+
+In this document, *GRLEN* is the bit width of general-purpose register, *FRLEN* is the bit width of float-point register and *WOA* is the bit width of the argument.
+The argument registers are collectively denoted as *AR* (*GAR* or *FAR*), where the general-purpose argument register is denoted as *GAR* and the floating-point argument register is denoted as *FAR*.
+
+=== ABI LP64D
+
+That is, *GRLEN* = 64, *FRLEN* = 64.
+
+=== C Data Types and Alignment
+
+The C data types and alignment in the LP64D ABI are defined in the table 3.
+
+The unsigned integer data types are zero-extended when stored in general-purpose register, and the signed integer data types are sign-extended.
+
+In the *LP64D* ABI, 32-bit types, such as `*int*`, are stored in general-purpose registers as proper sign extensions of their 32-bit values.
+This restriction holds even for unsigned 32-bit types.
+
+=== Argument passing
+
+Generally speaking, FARs are only used to pass floating-point arguments, GARs are used to pass non floating-point arguments and floating-point arguments when no FAR is available(long double type is also passed in a pair of GARs) and the reference.
+
+Arguments passed by reference may be modified by the callee.
+
+==== Scalar
+
+There are three cases:
+
+. 0 < WOA ≤ GRLEN
+
+.. Argument is passed in a single AR, or on the stack by value if none is available.
+
+... If the argument is float-point type, the argument is passed in FAR.
+if no FAR is available, it’s passed in GAR. If no GAR is available, it’s passed on the stack.
+When passed in registers or on the stack, floating-point types narrower than GRLEN bits are widened to GRLEN bits, with the upper bits undefined.
+
+... If the argument is integer or pointer type, the argument is passed in GAR.
+If no GAR is available, it’s passed on the stack.
+When passed in registers or on the stack, integer scalars narrower than GRLEN bits are sign-extended to GRLEN bits.
```

Comment:
Sign-extension typically refers to signed integers. Are unsigned integers zero extended, or treated sign-extended as if the integer was signed?

For example, is `uint8_t` value of `0xFF` passed in as `0x00000000000000FF` or as `0xFFFFFFFFFFFFFFFF` ?

Reply:
> Sign-extension typically refers to signed integers. Are unsigned integers zero extended, or treated sign-extended as if the integer was signed?
> 
> For example, is `uint8_t` value of `0xFF` passed in as `0x00000000000000FF` or as `0xFFFFFFFFFFFFFFFF` ?

both signed and unsigned are passed as `0xFFFFFFFFFFFFFFFF` .

Reply:
Is there a rationale behind this choice? I think that this design requires extra instructions to properly sign-extend and zero-extend the upper bits of unsigned integers. It just adds an extra overhead.

Reply:
> Is there a rationale behind this choice? 

This depends on the processor's architecture. Liking the LoongArch64, MIPS64, the most operations of 32-bits's instruction, liking compare, add, multiply.... , have striction on the high-32bits.


> I think that this design requires extra instructions to properly sign-extend and zero-extend the upper bits of unsigned integers. It just adds an extra overhead.

Reply:
I understand that it may be beneficial to widen arguments or return values given the processor architecture. I have seen it done in other ABIs.

However, it is very surprising to me to widen unsigned values via sign-extension as if they were signed. The other ABIs widen the unsigned values via zero-extensions. For example, RISC-V that you took a lot of inspiration from widens unsigned values via zero-extension as far as I can tell.

Reply:
Sorry for jumping in, but I think maybe @shushanhf should confirm with their teammates the actual implemented behavior in the gcc port. If the implementation is primarily derived from the RISC-V port, then such behavior is unlikely to be delibrately changed.